source/slang/slang-emit-llvm.cpp

```diff
@@ -0,0 +1,3127 @@
+    IRType* legalizeResourceTypes(IRType* type)
+    {
+        if (legalizedResourceTypeMap.containsKey(type))
+            return legalizedResourceTypeMap.getValue(type);
+
+        IRBuilder irBuilder(type->getModule());
+
+        IRType* legalizedType = type;
+        switch (type->getOp())
+        {
+        case kIROp_ConstantBufferType:
+        case kIROp_ParameterBlockType:
+            legalizedType = irBuilder.getRawPointerType();
+            break;
+        case kIROp_HLSLStructuredBufferType:
+        case kIROp_HLSLRWStructuredBufferType:
+        case kIROp_HLSLByteAddressBufferType:
+        case kIROp_HLSLRWByteAddressBufferType:
+            {
+                IRStructType* s = irBuilder.createStructType();
+                auto ptrKey = irBuilder.createStructKey();
+                auto sizeKey = irBuilder.createStructKey();
+                irBuilder.createStructField(s, ptrKey, irBuilder.getRawPointerType());
+                irBuilder.createStructField(s, sizeKey, irBuilder.getType(kIROp_UIntPtrType));
+                legalizedType = s;
+            }
+            break;
+        case kIROp_StructType:
+            {
+                auto structType = static_cast<IRStructType*>(type);
+                bool illegal = false;
+                for (auto field : structType->getFields())
+                {
+                    auto fieldType = field->getFieldType();
+                    auto legalizedFieldType = legalizeResourceTypes(fieldType);
+                    if (legalizedFieldType != fieldType)
+                        illegal = true;
+                }
+
+                if (illegal)
+                {
+                    IRStructType* s = irBuilder.createStructType();
+                    for (auto field : structType->getFields())
+                    {
+                        auto fieldType = field->getFieldType();
+                        auto legalizedFieldType = legalizeResourceTypes(fieldType);
+                        irBuilder.createStructField(s, field->getKey(), legalizedFieldType);
+                    }
+                    legalizedType = s;
+                }
+            }
+            break;
+        case kIROp_ArrayType:
+            {
+                auto arrayType = as<IRArrayTypeBase>(type);
+                auto elemType = arrayType->getElementType();
+                auto legalizedElemType = legalizeResourceTypes(elemType);
+                if (elemType != legalizedElemType)
+                    legalizedType =
+                        irBuilder.getArrayType(legalizedElemType, arrayType->getElementCount());
+            }
+            break;
+        default:
+            break;
+        }
+
+        legalizedResourceTypeMap[type] = legalizedType;
+        return legalizedType;
+    }
+
+    // Use this instead of the regular Slang::getOffset(), it handles
+    // legalized resource types correctly.
+    IRIntegerValue getOffset(IRStructField* field, IRTypeLayoutRules* rules)
```

Comment:
The comment here is another *strong* indication that the legalization being done here ought to be done upstream.

Reply:
Alright, I can add a separate type legalization pass. I think it should only be run for LLVM-on-CPU. The C++ target is already doing its own thing with the prelude that I don't want to disturb, and some future LLVM-on-GPU work would definitely want to lower resource types into specific builtins depending on target triple.